## Pandas 

pandas is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool,
built on top of the Python programming language.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("Glasgow_Weather.csv")

In [3]:
df

day  tempMin  tempMax                                  summary  \
0     2015-01-01     7.53    12.23       Possible light rain until evening.   
1     2015-01-02     3.58     7.15  Possible light rain throughout the day.   
2     2015-01-03    -0.61     6.54                Clear throughout the day.   
3     2015-01-04    -0.63     7.59        Mostly cloudy throughout the day.   
4     2015-01-05     6.51    10.43             Overcast throughout the day.   
...          ...      ...      ...                                      ...   
1790  2019-11-26     8.55    10.57        Mostly cloudy throughout the day.   
1791  2019-11-27     5.79     9.40        Mostly cloudy throughout the day.   
1792  2019-11-28     2.30     6.67        Mostly cloudy throughout the day.   
1793  2019-11-29     1.08     6.14        Mostly cloudy throughout the day.   
1794  2019-11-30     0.42     5.69        Mostly cloudy throughout the day.   

                   desc  cloudCover  humidity  windSpeed  visibility  
0                  rain        0.80      0.89      14.69        4.43  
1                  rain        0.62      0.79      15.04        5.64  
2                  rain        0.31      0.84       4.48        6.20  
3     partly-cloudy-day        0.78      0.85       4.35        6.22  
4     partly-cloudy-day        0.85      0.91       6.20        5.91  
...                 ...         ...       ...        ...         ...  
1790               rain        0.96      0.85       6.80       10.00  
1791               rain        0.87      0.85       6.82        9.03  
1792  partly-cloudy-day        0.40      0.79       5.15       10.00  
1793          clear-day        0.18      0.76       4.85        9.11  
1794  partly-cloudy-day        0.55      0.72       4.79       10.00  

[1795 rows x 9 columns]

You can find the dimension of your data base by using the attribute shape. Be aware it is an attribute because we do not use the rounded brackets at the end.

In [4]:
df.shape

(1795, 9)

The function info will give you greater detail on each column, stating number of non-null data points and the type of data.

If you see object, it normally means it is a string but if you aren't sure take a look at the data yourself.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1795 entries, 0 to 1794
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   day         1795 non-null   object 
 1   tempMin     1795 non-null   float64
 2   tempMax     1795 non-null   float64
 3   summary     1788 non-null   object 
 4   desc        1611 non-null   object 
 5   cloudCover  1792 non-null   float64
 6   humidity    1795 non-null   float64
 7   windSpeed   1795 non-null   float64
 8   visibility  1795 non-null   float64
dtypes: float64(6), object(3)
memory usage: 126.3+ KB


To view the data you can use the function head or tail. If you leave the brackets empty it shows only 5 rows but you can tell it how many to show. 

Be aware after 60 rows it will squash the rows together just like when you output the dataframe itself.

In [6]:
df.head()

day  tempMin  tempMax                                  summary  \
0  2015-01-01     7.53    12.23       Possible light rain until evening.   
1  2015-01-02     3.58     7.15  Possible light rain throughout the day.   
2  2015-01-03    -0.61     6.54                Clear throughout the day.   
3  2015-01-04    -0.63     7.59        Mostly cloudy throughout the day.   
4  2015-01-05     6.51    10.43             Overcast throughout the day.   

                desc  cloudCover  humidity  windSpeed  visibility  
0               rain        0.80      0.89      14.69        4.43  
1               rain        0.62      0.79      15.04        5.64  
2               rain        0.31      0.84       4.48        6.20  
3  partly-cloudy-day        0.78      0.85       4.35        6.22  
4  partly-cloudy-day        0.85      0.91       6.20        5.91

In [7]:
df.tail()

day  tempMin  tempMax                            summary  \
1790  2019-11-26     8.55    10.57  Mostly cloudy throughout the day.   
1791  2019-11-27     5.79     9.40  Mostly cloudy throughout the day.   
1792  2019-11-28     2.30     6.67  Mostly cloudy throughout the day.   
1793  2019-11-29     1.08     6.14  Mostly cloudy throughout the day.   
1794  2019-11-30     0.42     5.69  Mostly cloudy throughout the day.   

                   desc  cloudCover  humidity  windSpeed  visibility  
1790               rain        0.96      0.85       6.80       10.00  
1791               rain        0.87      0.85       6.82        9.03  
1792  partly-cloudy-day        0.40      0.79       5.15       10.00  
1793          clear-day        0.18      0.76       4.85        9.11  
1794  partly-cloudy-day        0.55      0.72       4.79       10.00

More indepth look at data can be done by using describe and correlation.

In [8]:
df.describe()

tempMin      tempMax   cloudCover     humidity    windSpeed  \
count  1795.000000  1795.000000  1792.000000  1795.000000  1795.000000   
mean      6.037253    12.998630     0.596964     0.818462     7.013616   
std       4.720846     5.270358     0.224010     0.076162     3.924601   
min      -8.670000    -1.390000     0.000000     0.500000     0.140000   
25%       2.590000     9.225000     0.450000     0.770000     4.115000   
50%       6.220000    13.050000     0.630000     0.830000     6.220000   
75%       9.800000    16.830000     0.770000     0.870000     9.135000   
max      18.030000    30.360000     0.990000     1.000000    24.840000   

        visibility  
count  1795.000000  
mean      5.773359  
std       0.797354  
min       0.190000  
25%       5.560000  
50%       6.030000  
75%       6.220000  
max      10.000000

In [9]:
df.corr()

tempMin   tempMax  cloudCover  humidity  windSpeed  visibility
tempMin     1.000000  0.804079    0.271800 -0.054032   0.187183    0.167162
tempMax     0.804079  1.000000   -0.114470 -0.362103   0.018456    0.174937
cloudCover  0.271800 -0.114470    1.000000  0.584773   0.300535   -0.263438
humidity   -0.054032 -0.362103    0.584773  1.000000  -0.016887   -0.429461
windSpeed   0.187183  0.018456    0.300535 -0.016887   1.000000    0.144079
visibility  0.167162  0.174937   -0.263438 -0.429461   0.144079    1.000000

You can filter on each column on the dataframe and a single column will be called a series.

Use the dataframe name and square brackets is how to filter your data.

In [10]:
df['tempMin']

0       7.53
1       3.58
2      -0.61
3      -0.63
4       6.51
        ... 
1790    8.55
1791    5.79
1792    2.30
1793    1.08
1794    0.42
Name: tempMin, Length: 1795, dtype: float64

Again the above will look familiar to you, as its almost like a dictionary.

If you want to view multiple columns at once, then replace the string with a list of the column names.

In [11]:
df[ ['tempMin', 'tempMax'] ]

tempMin  tempMax
0        7.53    12.23
1        3.58     7.15
2       -0.61     6.54
3       -0.63     7.59
4        6.51    10.43
...       ...      ...
1790     8.55    10.57
1791     5.79     9.40
1792     2.30     6.67
1793     1.08     6.14
1794     0.42     5.69

[1795 rows x 2 columns]

The order in which you type the list is the order in which the filtered dataframe will output.

In [12]:
df[['tempMax', 'tempMin']]

tempMax  tempMin
0       12.23     7.53
1        7.15     3.58
2        6.54    -0.61
3        7.59    -0.63
4       10.43     6.51
...       ...      ...
1790    10.57     8.55
1791     9.40     5.79
1792     6.67     2.30
1793     6.14     1.08
1794     5.69     0.42

[1795 rows x 2 columns]

To find out all the columns use the attribute columns. This is useful if you are working on a very large dataframe where the column names are squashed together when output.

In [13]:
df.columns

Index(['day', 'tempMin', 'tempMax', 'summary', 'desc', 'cloudCover',
       'humidity', 'windSpeed', 'visibility'],
      dtype='object')

If you are wanting to find the data for a particular row then you will use the iloc function. 

Image the iloc function like the list indexing. If you want the first row you pass 0 into the function.

In [14]:
df.iloc[0]

day                                   2015-01-01
tempMin                                     7.53
tempMax                                    12.23
summary       Possible light rain until evening.
desc                                        rain
cloudCover                                   0.8
humidity                                    0.89
windSpeed                                  14.69
visibility                                  4.43
Name: 0, dtype: object

If you want a single row, it provides a data series.

Just like if you want to slice a list use colon between the numbers in the iloc function. It will return a dataframe.

In [15]:
df.iloc[0:5]

day  tempMin  tempMax                                  summary  \
0  2015-01-01     7.53    12.23       Possible light rain until evening.   
1  2015-01-02     3.58     7.15  Possible light rain throughout the day.   
2  2015-01-03    -0.61     6.54                Clear throughout the day.   
3  2015-01-04    -0.63     7.59        Mostly cloudy throughout the day.   
4  2015-01-05     6.51    10.43             Overcast throughout the day.   

                desc  cloudCover  humidity  windSpeed  visibility  
0               rain        0.80      0.89      14.69        4.43  
1               rain        0.62      0.79      15.04        5.64  
2               rain        0.31      0.84       4.48        6.20  
3  partly-cloudy-day        0.78      0.85       4.35        6.22  
4  partly-cloudy-day        0.85      0.91       6.20        5.91

You can also pass a list of indexs in any order.

In [16]:
df.iloc[[0,2,4,3,5]]

day  tempMin  tempMax  \
0  2015-01-01     7.53    12.23   
2  2015-01-03    -0.61     6.54   
4  2015-01-05     6.51    10.43   
3  2015-01-04    -0.63     7.59   
5  2015-01-06     3.66     9.95   

                                             summary               desc  \
0                 Possible light rain until evening.               rain   
2                          Clear throughout the day.               rain   
4                       Overcast throughout the day.  partly-cloudy-day   
3                  Mostly cloudy throughout the day.  partly-cloudy-day   
5  Possible light rain in the morning and afternoon.               rain   

   cloudCover  humidity  windSpeed  visibility  
0        0.80      0.89      14.69        4.43  
2        0.31      0.84       4.48        6.20  
4        0.85      0.91       6.20        5.91  
3        0.78      0.85       4.35        6.22  
5        0.63      0.88       8.70        5.93

If you only want certain rows with certain columns then seperate the row definition with a comma and use the syntax as the rows.

You are passing 2 lists so remember that iloc will look like a nested list (extra square brackets). Additonally if you use colon notation, you will need to remove the brackets for the list.

In [17]:
df.iloc[[0,1],[0,1,2]]

day  tempMin  tempMax
0  2015-01-01     7.53    12.23
1  2015-01-02     3.58     7.15

In [18]:
df.iloc[0:2,0:3]

day  tempMin  tempMax
0  2015-01-01     7.53    12.23
1  2015-01-02     3.58     7.15

If your dataframe has an index you can still use iloc but you will find it difficult to find the index you are looking for. 

Pandas has the function loc[] which you pass in the unique index which will not be a integer to end of data rows.

In our current dataframe the index is numbers so we can still use loc with numbers.

In [19]:
df.loc[[0,1],["summary", "cloudCover"]]

summary  cloudCover
0       Possible light rain until evening.        0.80
1  Possible light rain throughout the day.        0.62

It makes most sense that the dataframe index would be the date, as an index should be unique, so we shall set the dataframes index to be "day"

In [20]:
df.set_index("day")

tempMin  tempMax                                  summary  \
day                                                                     
2015-01-01     7.53    12.23       Possible light rain until evening.   
2015-01-02     3.58     7.15  Possible light rain throughout the day.   
2015-01-03    -0.61     6.54                Clear throughout the day.   
2015-01-04    -0.63     7.59        Mostly cloudy throughout the day.   
2015-01-05     6.51    10.43             Overcast throughout the day.   
...             ...      ...                                      ...   
2019-11-26     8.55    10.57        Mostly cloudy throughout the day.   
2019-11-27     5.79     9.40        Mostly cloudy throughout the day.   
2019-11-28     2.30     6.67        Mostly cloudy throughout the day.   
2019-11-29     1.08     6.14        Mostly cloudy throughout the day.   
2019-11-30     0.42     5.69        Mostly cloudy throughout the day.   

                         desc  cloudCover  humidity  windSpeed  visibility  
day                                                                         
2015-01-01               rain        0.80      0.89      14.69        4.43  
2015-01-02               rain        0.62      0.79      15.04        5.64  
2015-01-03               rain        0.31      0.84       4.48        6.20  
2015-01-04  partly-cloudy-day        0.78      0.85       4.35        6.22  
2015-01-05  partly-cloudy-day        0.85      0.91       6.20        5.91  
...                       ...         ...       ...        ...         ...  
2019-11-26               rain        0.96      0.85       6.80       10.00  
2019-11-27               rain        0.87      0.85       6.82        9.03  
2019-11-28  partly-cloudy-day        0.40      0.79       5.15       10.00  
2019-11-29          clear-day        0.18      0.76       4.85        9.11  
2019-11-30  partly-cloudy-day        0.55      0.72       4.79       10.00  

[1795 rows x 8 columns]

In [21]:
df

day  tempMin  tempMax                                  summary  \
0     2015-01-01     7.53    12.23       Possible light rain until evening.   
1     2015-01-02     3.58     7.15  Possible light rain throughout the day.   
2     2015-01-03    -0.61     6.54                Clear throughout the day.   
3     2015-01-04    -0.63     7.59        Mostly cloudy throughout the day.   
4     2015-01-05     6.51    10.43             Overcast throughout the day.   
...          ...      ...      ...                                      ...   
1790  2019-11-26     8.55    10.57        Mostly cloudy throughout the day.   
1791  2019-11-27     5.79     9.40        Mostly cloudy throughout the day.   
1792  2019-11-28     2.30     6.67        Mostly cloudy throughout the day.   
1793  2019-11-29     1.08     6.14        Mostly cloudy throughout the day.   
1794  2019-11-30     0.42     5.69        Mostly cloudy throughout the day.   

                   desc  cloudCover  humidity  windSpeed  visibility  
0                  rain        0.80      0.89      14.69        4.43  
1                  rain        0.62      0.79      15.04        5.64  
2                  rain        0.31      0.84       4.48        6.20  
3     partly-cloudy-day        0.78      0.85       4.35        6.22  
4     partly-cloudy-day        0.85      0.91       6.20        5.91  
...                 ...         ...       ...        ...         ...  
1790               rain        0.96      0.85       6.80       10.00  
1791               rain        0.87      0.85       6.82        9.03  
1792  partly-cloudy-day        0.40      0.79       5.15       10.00  
1793          clear-day        0.18      0.76       4.85        9.11  
1794  partly-cloudy-day        0.55      0.72       4.79       10.00  

[1795 rows x 9 columns]

As you can see above the index still remain as the numbers from 0 to 1794. Pandas allows you to see what the result of your code is before it saves it. 

To save it you can either pass a paramenter of True to inplace or use assignment expression.

In [22]:
df.set_index("day", inplace=True)

In [23]:
df

tempMin  tempMax                                  summary  \
day                                                                     
2015-01-01     7.53    12.23       Possible light rain until evening.   
2015-01-02     3.58     7.15  Possible light rain throughout the day.   
2015-01-03    -0.61     6.54                Clear throughout the day.   
2015-01-04    -0.63     7.59        Mostly cloudy throughout the day.   
2015-01-05     6.51    10.43             Overcast throughout the day.   
...             ...      ...                                      ...   
2019-11-26     8.55    10.57        Mostly cloudy throughout the day.   
2019-11-27     5.79     9.40        Mostly cloudy throughout the day.   
2019-11-28     2.30     6.67        Mostly cloudy throughout the day.   
2019-11-29     1.08     6.14        Mostly cloudy throughout the day.   
2019-11-30     0.42     5.69        Mostly cloudy throughout the day.   

                         desc  cloudCover  humidity  windSpeed  visibility  
day                                                                         
2015-01-01               rain        0.80      0.89      14.69        4.43  
2015-01-02               rain        0.62      0.79      15.04        5.64  
2015-01-03               rain        0.31      0.84       4.48        6.20  
2015-01-04  partly-cloudy-day        0.78      0.85       4.35        6.22  
2015-01-05  partly-cloudy-day        0.85      0.91       6.20        5.91  
...                       ...         ...       ...        ...         ...  
2019-11-26               rain        0.96      0.85       6.80       10.00  
2019-11-27               rain        0.87      0.85       6.82        9.03  
2019-11-28  partly-cloudy-day        0.40      0.79       5.15       10.00  
2019-11-29          clear-day        0.18      0.76       4.85        9.11  
2019-11-30  partly-cloudy-day        0.55      0.72       4.79       10.00  

[1795 rows x 8 columns]

To reset the index simply use the reset_index and pass true to inplace to save the result. Again you can also use assignment to save the result.

In [24]:
df.reset_index(inplace=True)
df

day  tempMin  tempMax                                  summary  \
0     2015-01-01     7.53    12.23       Possible light rain until evening.   
1     2015-01-02     3.58     7.15  Possible light rain throughout the day.   
2     2015-01-03    -0.61     6.54                Clear throughout the day.   
3     2015-01-04    -0.63     7.59        Mostly cloudy throughout the day.   
4     2015-01-05     6.51    10.43             Overcast throughout the day.   
...          ...      ...      ...                                      ...   
1790  2019-11-26     8.55    10.57        Mostly cloudy throughout the day.   
1791  2019-11-27     5.79     9.40        Mostly cloudy throughout the day.   
1792  2019-11-28     2.30     6.67        Mostly cloudy throughout the day.   
1793  2019-11-29     1.08     6.14        Mostly cloudy throughout the day.   
1794  2019-11-30     0.42     5.69        Mostly cloudy throughout the day.   

                   desc  cloudCover  humidity  windSpeed  visibility  
0                  rain        0.80      0.89      14.69        4.43  
1                  rain        0.62      0.79      15.04        5.64  
2                  rain        0.31      0.84       4.48        6.20  
3     partly-cloudy-day        0.78      0.85       4.35        6.22  
4     partly-cloudy-day        0.85      0.91       6.20        5.91  
...                 ...         ...       ...        ...         ...  
1790               rain        0.96      0.85       6.80       10.00  
1791               rain        0.87      0.85       6.82        9.03  
1792  partly-cloudy-day        0.40      0.79       5.15       10.00  
1793          clear-day        0.18      0.76       4.85        9.11  
1794  partly-cloudy-day        0.55      0.72       4.79       10.00  

[1795 rows x 9 columns]

In [25]:
df = df.set_index("day")
df

tempMin  tempMax                                  summary  \
day                                                                     
2015-01-01     7.53    12.23       Possible light rain until evening.   
2015-01-02     3.58     7.15  Possible light rain throughout the day.   
2015-01-03    -0.61     6.54                Clear throughout the day.   
2015-01-04    -0.63     7.59        Mostly cloudy throughout the day.   
2015-01-05     6.51    10.43             Overcast throughout the day.   
...             ...      ...                                      ...   
2019-11-26     8.55    10.57        Mostly cloudy throughout the day.   
2019-11-27     5.79     9.40        Mostly cloudy throughout the day.   
2019-11-28     2.30     6.67        Mostly cloudy throughout the day.   
2019-11-29     1.08     6.14        Mostly cloudy throughout the day.   
2019-11-30     0.42     5.69        Mostly cloudy throughout the day.   

                         desc  cloudCover  humidity  windSpeed  visibility  
day                                                                         
2015-01-01               rain        0.80      0.89      14.69        4.43  
2015-01-02               rain        0.62      0.79      15.04        5.64  
2015-01-03               rain        0.31      0.84       4.48        6.20  
2015-01-04  partly-cloudy-day        0.78      0.85       4.35        6.22  
2015-01-05  partly-cloudy-day        0.85      0.91       6.20        5.91  
...                       ...         ...       ...        ...         ...  
2019-11-26               rain        0.96      0.85       6.80       10.00  
2019-11-27               rain        0.87      0.85       6.82        9.03  
2019-11-28  partly-cloudy-day        0.40      0.79       5.15       10.00  
2019-11-29          clear-day        0.18      0.76       4.85        9.11  
2019-11-30  partly-cloudy-day        0.55      0.72       4.79       10.00  

[1795 rows x 8 columns]

Now you can see using loc with the unique index of day, returns the result for that row.

In [26]:
df.loc["2019-11-30"]

tempMin                                    0.42
tempMax                                    5.69
summary       Mostly cloudy throughout the day.
desc                          partly-cloudy-day
cloudCover                                 0.55
humidity                                   0.72
windSpeed                                  4.79
visibility                                   10
Name: 2019-11-30, dtype: object

The functionality of loc isthe same as iloc, so you can slice both rows and columns. Using list and colons.

In [27]:
df.loc[["2019-11-30"],"summary":"cloudCover"]

summary               desc  cloudCover
day                                                                         
2019-11-30  Mostly cloudy throughout the day.  partly-cloudy-day        0.55

In [28]:
df.loc["2019-11-01":"2019-11-30","tempMax":"visibility"]

tempMax                                            summary  \
day                                                                      
2019-11-01    10.50                     Light rain throughout the day.   
2019-11-02    12.15                         Light rain in the morning.   
2019-11-03    11.80                 Possible drizzle in the afternoon.   
2019-11-04    10.62                 Possible light rain until evening.   
2019-11-05    10.33                                   Foggy overnight.   
2019-11-06     7.85        Possible drizzle starting in the afternoon.   
2019-11-07     8.18                   Possible drizzle in the morning.   
2019-11-08     7.81                          Clear throughout the day.   
2019-11-09     5.29                          Clear throughout the day.   
2019-11-10     5.72           Light rain in the evening and overnight.   
2019-11-11     7.38                     Light rain throughout the day.   
2019-11-12    10.76                   Possible drizzle in the morning.   
2019-11-13     5.94                Foggy in the morning and overnight.   
2019-11-14     6.49                          Clear throughout the day.   
2019-11-15     8.69                  Partly cloudy throughout the day.   
2019-11-16     4.63                          Clear throughout the day.   
2019-11-17     6.45                                   Foggy overnight.   
2019-11-18     4.73                Foggy in the morning and overnight.   
2019-11-19     3.67  Possible light rain in the evening and overnight.   
2019-11-20     7.40                   Possible drizzle in the morning.   
2019-11-21     7.78                  Mostly cloudy throughout the day.   
2019-11-22     8.91                  Mostly cloudy throughout the day.   
2019-11-23     9.38                  Mostly cloudy throughout the day.   
2019-11-24    10.02                  Mostly cloudy throughout the day.   
2019-11-25    10.45                  Mostly cloudy throughout the day.   
2019-11-26    10.57                  Mostly cloudy throughout the day.   
2019-11-27     9.40                  Mostly cloudy throughout the day.   
2019-11-28     6.67                  Mostly cloudy throughout the day.   
2019-11-29     6.14                  Mostly cloudy throughout the day.   
2019-11-30     5.69                  Mostly cloudy throughout the day.   

                         desc  cloudCover  humidity  windSpeed  visibility  
day                                                                         
2019-11-01               rain        0.97      0.96       7.48        2.00  
2019-11-02               rain        0.76      0.95       3.76        6.21  
2019-11-03               rain        0.73      0.88       7.97        6.07  
2019-11-04               rain        0.70      0.87      13.51        5.42  
2019-11-05  partly-cloudy-day        0.74      0.77       6.50        6.48  
2019-11-06               rain        0.61      0.88       5.05        5.74  
2019-11-07               rain        0.68      0.84      12.76        5.81  
2019-11-08          clear-day        0.21      0.77       3.14        6.22  
2019-11-09  partly-cloudy-day        0.34      0.83       7.07        6.16  
2019-11-10               rain        0.32      0.82       4.85        6.10  
2019-11-11               rain        0.78      0.87      11.88        5.98  
2019-11-12               rain        0.43      0.85       6.42        5.93  
2019-11-13  partly-cloudy-day        0.62      0.92       4.95        5.51  
2019-11-14          clear-day        0.26      0.76       4.55        5.55  
2019-11-15  partly-cloudy-day        0.40      0.76       4.64        6.34  
2019-11-16          clear-day        0.27      0.93       1.70        6.01  
2019-11-17          clear-day        0.30      0.94       0.92        5.60  
2019-11-18  partly-cloudy-day        0.35      0.93       0.86        3.70  
2019-11-19               rain        0.69      0.89       5.03        4.58  
2019-11-20               r